### PR TITLE
TorchWatcher: Track layer-wise metrics from PyTorch models to Neptune

### PR DESCRIPTION
**Pull request by LeoRoccoBreedt**
# Description

Include a summary of the changes and the related issue.

__Related to:__ <ClickUp/JIRA task name>

__Any expected test failures?__

---

Add a `[X]` to relevant checklist items

## ❔ This change

- [ ] adds a new feature
- [ ] fixes breaking code
- [ ] is cosmetic (refactoring/reformatting)

---

## ✔️ Pre-merge checklist

- [ ] Refactored code ([sourcery](https://sourcery.ai/))
- [ ] Tested code locally
- [ ] Precommit installed and run before pushing changes
- [ ] Added code to GitHub tests ([notebooks](workflows/test-notebooks.yml), [scripts](workflows/test-scripts.yml))
- [ ] Updated GitHub [README](../README.md)
- [ ] Updated the projects overview page on Notion

---

## 🧪 Test Configuration

- OS:
- Python version:
- Neptune version:
- Affected libraries with version:

### CODE REVIEW

**Comment by SiddhantSadangi** [comment]
@LeoRoccoBreedt - is this ready for review?

File: utils/monitoring_tools/pytorch_model_internals/scripts/TorchWatcher.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+        if track_layers is not None:
+            invalid_layers = [layer for layer in track_layers if layer not in PYTORCH_LAYERS]
+            if invalid_layers:
+                raise ValueError(
+                    f"Invalid layer types specified: {invalid_layers}. "
+                    f"Please use valid PyTorch layer types from torch.nn."
+                )
+
+        self.model = model
+        self.hooks: List[torch.utils.hooks.RemovableHandle] = []
```
[the remark this answers is not in the document]
Comment (LeoRoccoBreedt) [reply utils/monitoring_tools/pytorch_model_internals/scripts/TorchWatcher.py:114]: Can you provide a suggested implementation?

[the remark this answers is not in the document]
Comment (LeoRoccoBreedt) [reply utils/monitoring_tools/pytorch_model_internals/scripts/TorchWatcher.py:114]: What would be the chances that this implementation from PyTorch changes? Would it not be better to have the hooks be type annotated to the correct type rather than Any?

File: utils/monitoring_tools/pytorch_model_internals/scripts/TorchWatcher.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    A comprehensive tracking mechanism for PyTorch models with enhanced logging.
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        run: Any,  # Made more flexible to support different logging mechanisms
+        track_layers: Optional[List[Type[nn.Module]]] = None,
+        tensor_stats: Optional[List[str]] = None,
+        base_namespace: str = "debug",  # Default namespace for all metrics
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/scripts/TorchWatcher.py:217]: `model_internals` instead of `debug`? `debug` sounds a bit vague

File: utils/monitoring_tools/pytorch_model_internals/scripts/torch_watcher_example.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    # Create model and watcher
+    model = SimpleNet()
+
+    # Initialize watcher with specific layer types to track
+    watcher = TorchWatcher(
+        model,
+        run,
+        track_layers=[nn.Linear, nn.ReLU],  # Track only Linear and ReLU layers
+        tensor_stats=["mean", "norm"],  # Track mean and norm statistics
+        base_namespace="model_metrics",  # Default namespace for all metrics
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/scripts/torch_watcher_example.py:127]: `model_metrics` can also include evaluation metrics like `acc`, `loss`, etc.

```suggestion
        base_namespace="model_internals",  # Default namespace for all metrics
```

File: utils/monitoring_tools/pytorch_model_internals/scripts/TorchWatcher.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+            }
+            self._track_metric("parameters", parameters, namespace)
+
+    def watch(
+        self,
+        step: Union[int, float],
+        track_gradients: bool = True,
+        track_parameters: bool = True,
+        track_activations: bool = True,
+        namespace: Optional[str] = None,
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/scripts/TorchWatcher.py:329]: If this is acting as a `prefix` to the `base_namespace`, it should be named as such to make the information architeture clear

```suggestion
        prefix: Optional[str] = None,
```

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+- **Flexible Layer Selection**: Track all layers or specify which layer types to monitor
+- **Comprehensive Statistics**: Predefined tensor statistics including mean, standard deviation, norm, min, max, variance, and absolute mean
+- **Configurable Tracking**: Enable/disable tracking of activations, gradients, and parameters as needed
+- **Organized Logging**: Structured metric namespacing for better organization in Neptune
+- **Memory Efficient**: Clears stored tensors after each logging step
+- **Error Handling**: Robust error handling with informative warnings
+
+## Changelog
+
+### [v0.1.0] - 2024-03-19
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:29]: Date to be updated

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 6]
```diff
@@ -0,0 +1,252 @@
+# TorchWatcher
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:1]:
```suggestion
# Neptune TorchWatcher
```

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    - As model size increases, extracting model parameters causes increased time in training loop compared to gradients and activations which are extracted with PyTorch hooks.
+    - Larger models have less total overhead between model training time in batch vs. time taken to extract values and is minimal if logging activations and gradients. 
+- Benchmarking also showed that the average running time per batch remained constant, indicating that there is no leakage or slowdown between training batches.
+
+![Benchmark analysis](benchmark_analysis.png)
+*Figure 1: Training time overhead comparison between baseline training and training with TorchWatcher enabled. Results shown for different model sizes and tracking configurations.*
+
+### Future benchmarking
+- Analyze memory overhead
+- Train on multiple GPU's
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:239]:
```suggestion
- Train on multiple GPUs
```

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+## Features
+
+- **Automatic Layer Tracking**: Monitors activations, gradients, and parameters of specified PyTorch layers
+- **Flexible Layer Selection**: Track all layers or specify which layer types to monitor
+- **Comprehensive Statistics**: Predefined tensor statistics including mean, standard deviation, norm, min, max, variance, and absolute mean
+- **Configurable Tracking**: Enable/disable tracking of activations, gradients, and parameters as needed
+- **Organized Logging**: Structured metric namespacing for better organization in Neptune
+- **Memory Efficient**: Clears stored tensors after each logging step
+- **Error Handling**: Robust error handling with informative warnings
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:25]: Some of the points are repetitive

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+### [v0.1.0] - 2024-03-19
+
+#### Added
+- Initial release of TorchWatcher
+
+## Prerequisites
+
+- Python 3.8 or higher
+- PyTorch 2.0 or higher
+- CUDA 11.7 or higher (for GPU support)
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:38]: Tested on the minimum versions?

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+- Neptune account and API token
+
+## Installation
+
+```bash
+pip install neptune_scale
+```
+
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:46]: A bit misleading as it implies that installing `neptune_scale` will also install the `TorchWatcher`.
Having the `neptune_scale` package installed should be mentioned in the prerequisites

```suggestion
```

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+## Usage
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:47]: Mention that the `torchwatcher.py` needs to be placed in the CWD

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+- Convolutional layers
+- Recurrent layers
+- Normalization layers
+- Activation layers
+- Pooling layers
+- Dropout layers
+- Embedding layers
+- Transformer layers
+- Attention layers
+- And more...
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:192]:
```suggestion
```

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+
+A reproduction script can be found called - `benchmark_torchwatcher.py`.
+
+### Performance Impact
+
+TorchWatcher is designed to be lightweight and efficient. Our benchmarks show minimal impact on training batch time while providing comprehensive monitoring capabilities across varying model sizes. The largest contributor to performance degradation is the extraction of the model's named parameters (weights and biases). These are extracted using `model.named_parameters()` and creates increased overhead compared to the hooks that PyTorch models support natively for activations and gradients. Additional profiling showed that the `track_parameters` method created the most overhead and can be explored further for future package optimization.
+
+#### Analysis summary
+- Larger models (increase in parameters) have a linearly increasing cumulative time per epoch. 
+- Larger models spend more time in training loop per batch, as expected.
```
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:228]: These have nothing to do with TorchWatcher

```suggestion
```

File: utils/monitoring_tools/pytorch_model_internals/README.md (hunk 7 above)
Comment (SiddhantSadangi) [inline utils/monitoring_tools/pytorch_model_internals/README.md:239]:
```suggestion
- Train on multiple GPUs
```

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+- A single RTX5000 GPU in an isolated cluster
+- Various model architecture size with Linear and Relu layers only
+- Multiple tracking configurations for TorchWatcher
+- Generic numeric dataset
+- Training parameters:
+    - Samples: 4096
+    - Batch size: 512
+    - Epochs: 20
+
+A reproduction script can be found called - `benchmark_torchwatcher.py`.
```
[the remark this answers is not in the document]
Comment (SiddhantSadangi) [reply utils/monitoring_tools/pytorch_model_internals/README.md:194]: not really needed... we can share on-demand

**Review by SiddhantSadangi — APPROVED** [review]

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 15]
```diff
@@ -0,0 +1,236 @@
+# Neptune TorchWatcher
```
Comment (normandy7) [inline utils/monitoring_tools/pytorch_model_internals/README.md:1]: Is this an invented name for a utility we're developing and providing?

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+
+[![Explore in Neptune][Explore in Neptune badge]][Neptune dashboard]
+
+A lightweight PyTorch model monitoring tool that automatically tracks layer activations, gradients, and parameters during training. Built for seamless integration with neptune.ai.
+
+## Why TorchWatcher?
```
Comment (normandy7) [inline utils/monitoring_tools/pytorch_model_internals/README.md:7]: I honestly think the tone is off for this section. It sounds like marketing of a product; I first thought TorchWatcher is an external library that we integrate with.

Can we please drop the promotional vibe and focus on describing the tool in a neutral and factual way?  As long as the tool lives entirely within the examples repo, it should be clear that it's more of an "add-on", like the hardware monitoring utility.

→ Drop/change the heading and remove buzzwords

(The code itself can remain as-is)

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+### Quickstart
+
+See `scripts/torch_watcher_example.py` for a complete example demonstrating:
+- Model definition
+- Data generation
+- Training loop with different tracking configurations
+- Namespace organization
+- Integration with Neptune
+
+### Basic Usage
```
Comment (normandy7) [inline utils/monitoring_tools/pytorch_model_internals/README.md:43]: Sentence case for headings:

```suggestion
### Basic usage
```

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+- **Robust error handling**: Informative warnings and graceful error recovery
+
+Whether you're debugging gradient flow issues, monitoring activation patterns, or tracking parameter changes during training, TorchWatcher provides the insights you need without the overhead of manual implementation.
+
+## Changelog
+
+### [v0.1.0] - 2025-07-16
+
+#### Added
+- Initial release of TorchWatcher
```
Comment (normandy7) [inline utils/monitoring_tools/pytorch_model_internals/README.md:25]: Not a product

```suggestion
- Initial release
```

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+    step=current_step,
+    prefix="train"  # Metrics will be under "train/model_internals/..."
+)
+```
+
+### Namespace Organization
+
+TorchWatcher provides a hierarchical namespace structure for organizing metrics:
+
+1. **Base Namespace**: Set during initialization
```
Comment (normandy7) [inline utils/monitoring_tools/pytorch_model_internals/README.md:106]: Sentence case also for these

```suggestion
1. **Base namespace**: Set during initialization
```

File: utils/monitoring_tools/pytorch_model_internals/README.md [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+```
+
+Example metric names:
+- `train/model_internals/activations/fc1/mean`
+- `validation/model_internals/gradients/fc2/norm`
+- `train/model_internals/parameters/fc1/weight/mean`
+
+### Example Use Cases
+
+1. **Training with Full Tracking**:
```
Comment (normandy7) [inline utils/monitoring_tools/pytorch_model_internals/README.md:136]: Sentence case for these as well (and no bolding needed)

```suggestion
1. Training with full tracking:
```

**Review by normandy7 — CHANGES_REQUESTED** [review]

**Review by normandy7 — APPROVED** [review]
Looks good, thanks!

**Review by SiddhantSadangi — APPROVED** [review]
